service-commands/src/commands/seed/SeedSession.js

```diff
@@ -0,0 +1,106 @@
+  getRandomEmail,
+  getRandomPassword,
+  getRandomUserMetadata,
+} = RandomUtils
+
+class SeedSession {
+  constructor() {
+    this.cache = new UserCache()
+    this.localstorage = new LocalStorageWrapper()
+  }
+
+  init = async (libsConfigOverride = {}) => {
+      const libsConfig = getLibsConfig(libsConfigOverride)
+      this.libs = new AudiusLibs(libsConfig)
+      await this.libs.init()
+  }
+
+  clearSession = async () => {
+      this.localstorage.clear()
+      this.libs = undefined
+      this.cache.clear()
+  }
+
+  setUser = async ({ alias = '', userId = null }) => {
+      const userDetails = this.cache.findUser({ alias, userId })
+      const { hedgehogEntropyKey, userAlias } = userDetails
+      if (!userDetails.hedgehogEntropyKey) {
+        console.error(`No user with alias ${alias} or id ${userId} found in local seed cache.`)
+      } else {
+        this.localstorage.setUserEntropy(hedgehogEntropyKey)
+        this.cache.setActiveUser(userAlias)
+      }
+      const libsConfigOverride = {
+        identityServiceConfig: {
+          useHedgehogLocalStorage: true
+        }
+      }
+      await this.init(libsConfigOverride)
+      if (!this.libs.userStateManager.getCurrentUserId() === userDetails.userId) {
+        throw new Error(`Error calling SeedSession.setUser with alias ${alias} / userId ${userId} -- please check your seed cache at ${SEED_CACHE_PATH} to ensure that user exists.`)
+      }
+  }
+
+  createUser = async (alias, options) => {
+    let { email, password, metadata } = options
+    if (!email) {
+      email = getRandomEmail()
+    }
+    if (!password) {
+      password = getRandomPassword()
+    }
+    const randomMetadata = getRandomUserMetadata(email, password)
+    if (!metadata) {
+      metadata = randomMetadata
+    } else {
+      metadata = Object.assign(randomMetadata, metadata)
+    }
+    let signUpResponse
+    const profilePictureFile = null
+    const coverPhotoFile = null
+    const hasWallet = false
+    const host = (typeof window !== 'undefined' && window.location.origin) || null
+    const createWAudioUserBank = false
+    try {
+      signUpResponse = await this.libs.Account.signUp(email, password, metadata, profilePictureFile, coverPhotoFile, hasWallet, host, createWAudioUserBank)
+    } catch (error) {
+      console.log(error, signUpResponse)
+    }
+    if (signUpResponse.error) {
+      throw new Error(signUpResponse.error)
+    } else {
+      const hedgehogEntropyKey = this.localstorage.getUserEntropy()
+      const userId = signUpResponse.userId
+      if (!alias) {
+        alias = hedgehogEntropyKey
+      }
+      this.cache.addUser({ alias, hedgehogEntropyKey, userId })
+      this.cache.addLoginDetails({ entropy: hedgehogEntropyKey, email, password })
+      this.cache.setActiveUser(alias)
+      console.log(`Successfully seeded user with id: ${userId} and alias: ${alias}`)
+      return
+    }
+  }
+  // TODO API methods that call through to CLI things (wrapper layer)?
```

Comment:
Thinking about using a metaprogramming approach to expose the rest of the CLI methods on the SeedSession class, so that you can write scripts in Node as well as CLI scripts to compose seed sessions. I think this would be a cool way to start separating out some of the seed-specific logic from the CLI user-input parsing/sanitizing logic, but want to make sure there is need for this first. The intention would be a lightweight general wrapper (hence the metaprogramming approach to derive methods instead of declaring each one in code). Welcome any feedback here!

Reply:
Got a q from Marcus on this so adding this here to clarify - some examples!
Right now if you want to compose seed stuff you can write a script in the CLI that for instance creates an array or goes through a JSON and just applies each one to the seed cli commands via xargs or something or a loop - like
```js
# seed-setup
A seed create-user -n 10
A seed upload-track -u 5
for i in {1..10}; do
    A seed upload-track -u i 
    A seed follow-user -u i -f 2
    A seed create-playlist -u i -t 1
    A seed repost-track -u i -t 1
done
```
but you cant do this in Node -
```
// seed-setup.js
const { SeedSession } = require('path/to/seed')

for (let i in _.range(0, 10)) {
    let seed = new SeedSession()
    const user = seed.createUser()
    const userId = user.userId
    seed.uploadTrack({ user: userId })
    seed.createPlaylist({ user: userId, tracks: [1] })
    seed.repostTrack() // etc
}
```
I guess you could technically do it now by making those calls reach through to `seed.libs.Track`  etc and calling via `seed.libs.Track.uploadTrack` etc. This is doable and for default args for random stuff you could import in the `SeedUtils` functions. But maybe that’s a lot of work and we want to call it directly the simple way in `seed-setup.js` above and let seed take care of all the randomization etc.

if it’s useful to expose those as `SeedSession.uploadTrack` without making the user have to call through to `seed.libs.Track.uploadTrack` etc, but at the same time not create a whole other libs wrapper layer in service-commands, w metaprogramming, the idea would be to go from the CLI map and decorate the SeedSession class with additional methods - dynamically exposing those methods without needing to define each one on the actual class.

Does that make more sense? It’s like auto-generated node wrapper functions exposed on the SeedSession class.

Reply:
I think I follow - having node programmatic access to seed could probably be useful for mad-dog, though I'm not sure, without thinking more deeply about it, if it would be that much more useful than calling some of the service commands directly. I support shipping what we have here now, which looks great, using it a bit and seeing how our needs evolve!